README.rst
=============
ovchipcardlib
=============

A library to interact with the ov-chip card service.

It utilizes the webapi of the mobile apps to get the information.
Very loosely based on https://github.com/OVChip/ovchipapi-python for the
authentication part.

Be aware that the ov-chip card system used in the Netherlands for public
transport was not designed to be real time. So expect a hefty time drift
before the actual checkin/out action and information showing up on the api
and by extension on the transactions list.

Also the sequence of transactions is not guaranteed. It is absolutelly
certain that if there are a lot of checkins/outs closely together they will
appear out of order on the web api. Apparently there is a serverside job that
sorts them once a day.

The library handles that by sorting the list of transactions always. So on
any action it would return a properly ordered list of events but in case of
monitoring through the card.get_latest_transactions() method it is
absolutelly guaranteed that transaction will appear out of order and with a
delay ranging between 2 and 15 minutes according to some simple benchmarks.


* Documentation: http://ovchipcardlib.readthedocs.io/en/latest/

Features
--------

* Can get cards bound with the account and lists of transactions for the cards.


HISTORY.rst
.. :changelog:

History
-------

0.1 (14-04-2017)
---------------------

* First release on pypi


0.1.1 (07-07-2021)
------------------

* Updated template, bumped dependencies and implemented a pipeline.


tests/__init__.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# File: __init__.py
#
# Copyright 2017 Costas Tyfoxylos
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to
#  deal in the Software without restriction, including without limitation the
#  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
#  sell copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
#

"""
.. _Google Python Style Guide:
   http://google.github.io/styleguide/pyguide.html
"""

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = '''google'''
__date__ = '''14-04-2017'''
__copyright__ = '''Copyright 2017, Costas Tyfoxylos'''
__license__ = '''MIT'''
__maintainer__ = '''Costas Tyfoxylos'''
__email__ = '''<costas.tyf@gmail.com>'''
__status__ = '''Development'''  # "Prototype", "Development", "Production".


tests/test_ovchipcardlib.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# File: test_ovchipcardlib.py
#
# Copyright 2017 Costas Tyfoxylos
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to
#  deal in the Software without restriction, including without limitation the
#  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
#  sell copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
#

"""
test_ovchipcardlib
----------------------------------
Tests for `ovchipcardlib` module.

.. _Google Python Style Guide:
   http://google.github.io/styleguide/pyguide.html

"""

from betamax.fixtures import unittest

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = '''google'''
__date__ = '''14-04-2017'''
__copyright__ = '''Copyright 2017, Costas Tyfoxylos'''
__credits__ = ["Costas Tyfoxylos"]
__license__ = '''MIT'''
__maintainer__ = '''Costas Tyfoxylos'''
__email__ = '''<costas.tyf@gmail.com>'''
__status__ = '''Development'''  # "Prototype", "Development", "Production".


class TestOvchipcardlib(unittest.BetamaxTestCase):

    def setUp(self):
        """
        Test set up

        This is where you can setup things that you use throughout the tests. This method is called before every test.
        """
        pass

    def tearDown(self):
        """
        Test tear down

        This is where you should tear down what you've setup in setUp before. This method is called after every test.
        """
        pass


ovchipcardlib/helpers.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
# File: helpers.py
#
# Copyright 2017 Costas Tyfoxylos
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to
#  deal in the Software without restriction, including without limitation the
#  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
#  sell copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
#

"""
Main code for helpers.

.. _Google Python Style Guide:
   http://google.github.io/styleguide/pyguide.html

"""

import logging
from collections import namedtuple
from datetime import datetime
from time import sleep

import requests

from .ovchipcardlibexceptions import BrokenResponse

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = 'plaintext'
__date__ = '''31-03-2017'''

# This is the main prefix used for logging
LOGGER_BASENAME = '''ovchipcardhelpers'''
LOGGER = logging.getLogger(LOGGER_BASENAME)
LOGGER.addHandler(logging.NullHandler())


Token = namedtuple('Token', ('access',
                             'expires_in',
                             'id',
                             'refresh',
                             'scope',
                             'type'))


class Card:
    """Model of the card object."""

    def __init__(self, service_instance, data):
        logger_name = '{base}.{suffix}'.format(base=LOGGER_BASENAME,
                                               suffix=self.__class__.__name__)
        self._logger = logging.getLogger(logger_name)
        self._service = service_instance
        self._data = data
        self._transactions = []
        self._latest_transactions = []

    @property
    def alias(self):
        """Alias."""
        return self._data.get('alias')

    @property
    def auto_reloading(self):
        """Auto reloading."""
        return self._data.get('autoReloadEnabled')

    @property
    def balance(self):
        """Balance."""
        return float(self._data.get('balance')) / 100

    @property
    def balance_datetime(self):
        """Balance Datetime."""
        return self.timestamp_to_datetime(self._data.get('balanceDate'))

    @property
    def expiry_datetime(self):
        """Expiry datetime."""
        return self.timestamp_to_datetime(self._data.get('expiryDate'))

    @staticmethod
    def timestamp_to_datetime(timestamp):
        """Timestamt to datetime."""
        return datetime.fromtimestamp(float(timestamp / 1000))

    @property
    def default(self):
        """Default."""
        return self._data.get('defaultCard')

    @property
    def number(self):
        """Number."""
        return self._data.get('mediumId')

    @property
    def status(self):
        """Status."""
        return self._data.get('status')

    @property
    def status_announcement(self):
        """Status announcement."""
        return self._data.get('statusAnnouncement')

    @property
    def type(self):
        """Type."""
        return self._data.get('type')

    def get_latest_transactions(self):
        """Gets the latest transactions.

        Returns transactions happening after the last time this was called
        and clears itself upon conclusion. Can be used to monitor the live
        transactions on the card.

        Return:
             transactions (list): A list of transaction objects.

        """
        _ = self.transactions  # noqa
        transactions = self._latest_transactions[:]
        self._latest_transactions = []
        return transactions

    @property
    def transactions(self):
        """Presents the transactions of the card.

        Returns:
            transactions (list): A list of transaction objects.

        """
        if not self._transactions:  # pylint: disable=too-many-nested-blocks
            for transactions in self._get_transactions_by_window():
                self._transactions.extend(transactions)
                self._latest_transactions.extend(transactions)
        else:
            found = 0
            try:
                for transactions in self._get_transactions_by_window():
                    for transaction in transactions:
                        exists = next((trip for trip in self._transactions
                                       if trip.datetime == transaction.datetime
                                       ), None)
                        if exists:
                            found += 1
                            message = ('Transaction already '
                                       'exists {}').format(transaction)
                            self._logger.debug(message)
                            if found == 10:
                                self._logger.debug('Matched already 10 existing'
                                                   ' transactions, breaking...')
                                raise StopIteration
                        else:
                            self._transactions.append(transaction)
                            self._latest_transactions.append(transaction)
                            message = 'Added transaction {}'.format(transaction)
                            self._logger.debug(message)
                    # added a slight delay on retrieving from the api so it
                    # is not overwhelmed
                    sleep(0.5)
            except StopIteration:
                pass
        self._transactions.sort(key=lambda x: x.datetime, reverse=True)
        return self._transactions

    def _get_transactions_by_window(self):
        window = 20
        data = {}
        params = {'authorizationToken': self._service.authorization_token,
                  'mediumId': self.number,
                  'offset': 0,
                  'locale': self._service.locale}
        while not window > data.get('totalSize', 20):
            _, _, data = self._get_transaction_response(params)
            params['offset'] = int(data.get('nextOffset'))
            transactions = [Transaction(record)
                            for record in data.get('records')]
            yield transactions
        _, _, data = self._get_transaction_response(params)
        params['offset'] = int(data.get('nextOffset'))
        transactions = [Transaction(record)
                        for record in data.get('records')]
        yield transactions

    def _get_transaction_response(self, params):
        url = '{base}/transaction/list'.format(base=self._service.api_url)
        response = requests.post(url, data=params)
        try:
            data = response.json()
        except ValueError:
            message = 'Server Error. Response {}'.format(response.text)  #
            self._logger.debug(message)
            raise BrokenResponse(message)
        status_code = data.get('c')
        exception = data.get('e')
        data = data.get('o')
        if not status_code == 200:
            raise LookupError(exception or data)
        return status_code, exception, data


class Transaction:
    """Models the transactions."""

    def __init__(self, data):
        self._data = data
        self.timestamp_to_datetime = Card.timestamp_to_datetime

    def __str__(self):
        """String representation of the Transaction object."""
        text = ('Transaction type :{}'.format(self.name),
                'Starting point   :{}'.format(self.check_in_info),
                'Station          :{}'.format(self.station),
                'Price            :{}'.format(self.fare),
                'Date             :{}'.format(self.datetime),
                'Company Used     :{}'.format(self.company))
        return '\n'.join(text)

    @property
    def check_in_info(self):
        """Check in info."""
        return self._data.get('checkInInfo')

    @property
    def check_in_text(self):
        """Check in text."""
        return self._data.get('checkInText')

    @property
    def e_purse_mut(self):
        """E Purse mut."""
        return self._data.get('ePurseMut')

    @property
    def e_purse_mut_info(self):
        """E Purse mut info."""
        return self._data.get('ePurseMutInfo')

    @property
    def fare(self):
        """Fare."""
        return self._data.get('fare')

    @property
    def fare_calculation(self):
        """Fare calculation."""
        return self._data.get('fareCalculation')

    @property
    def fare_text(self):
        """Fare text."""
        return self._data.get('fareText')

    @property
    def type(self):
        """Type."""
        return self._data.get('modalType')

    @property
    def product_info(self):
        """Product Info."""
        return self._data.get('productInfo')

    @property
    def company(self):
        """Company."""
        return self._data.get('pto')

    @property
    def product_text(self):
        """Product Text."""
        return self._data.get('productText')

    @property
    def datetime(self):
        """Datetime."""
        return self.timestamp_to_datetime(self._data.get('transactionDateTime'))

    @property
    def explanation(self):
        """Explanation."""
        return self._data.get('transactionExplanation')

    @property
    def station(self):
        """Station."""
        return self._data.get('transactionInfo')

    @property
    def name(self):
        """Name."""
        return self._data.get('transactionName')

    @property
    def priority(self):
        """Priority."""
        return self._data.get('transactionPriority')


ovchipcardlib/__init__.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# File: __init__.py
#
# Copyright 2017 Costas Tyfoxylos
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to
#  deal in the Software without restriction, including without limitation the
#  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
#  sell copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
#

"""
ovchipcardlib package.

Import all parts from ovchipcardlib here

.. _Google Python Style Guide:
   http://google.github.io/styleguide/pyguide.html
"""
from ._version import __version__
from .helpers import Card, Token
from .ovchipcardlibexceptions import (UnableToRetrieveToken,
                                      UnAuthorized,
                                      BrokenResponse)
from .ovchipcardlib import Service

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = '''google'''
__date__ = '''14-04-2017'''
__copyright__ = '''Copyright 2017, Costas Tyfoxylos'''
__license__ = '''MIT'''
__maintainer__ = '''Costas Tyfoxylos'''
__email__ = '''<costas.tyf@gmail.com>'''
__status__ = '''Development'''  # "Prototype", "Development", "Production".

# This is to 'use' the module(s), so lint doesn't complain
assert __version__

# assert exceptions
assert UnableToRetrieveToken
assert UnAuthorized
assert BrokenResponse

# assert object
assert Card
assert Token
assert Service


_build/html/searchindex.js
Search.setIndex({envversion:46,filenames:["authors","contributing","history","index","installation","modules","ovchipcardlib","readme","usage"],objects:{"":{ovchipcardlib:[6,0,0,"-"]},"ovchipcardlib.helpers":{Card:[6,4,1,""],Token:[6,4,1,""],Transaction:[6,4,1,""]},"ovchipcardlib.helpers.Card":{"default":[6,1,1,""],alias:[6,1,1,""],auto_reloading:[6,1,1,""],balance:[6,1,1,""],balance_datetime:[6,1,1,""],expiry_datetime:[6,1,1,""],get_latest_transactions:[6,2,1,""],number:[6,1,1,""],status:[6,1,1,""],status_announcement:[6,1,1,""],timestamp_to_datetime:[6,5,1,""],transactions:[6,1,1,""],type:[6,1,1,""]},"ovchipcardlib.helpers.Token":{"__getnewargs__":[6,2,1,""],"__getstate__":[6,2,1,""],"__repr__":[6,2,1,""],access:[6,1,1,""],expires_in:[6,1,1,""],id:[6,1,1,""],refresh:[6,1,1,""],scope:[6,1,1,""],type:[6,1,1,""]},"ovchipcardlib.helpers.Transaction":{"__str__":[6,2,1,""],check_in_info:[6,1,1,""],check_in_text:[6,1,1,""],company:[6,1,1,""],datetime:[6,1,1,""],e_purse_mut:[6,1,1,""],e_purse_mut_info:[6,1,1,""],explanation:[6,1,1,""],fare:[6,1,1,""],fare_calculation:[6,1,1,""],fare_text:[6,1,1,""],name:[6,1,1,""],priority:[6,1,1,""],product_info:[6,1,1,""],product_text:[6,1,1,""],station:[6,1,1,""],type:[6,1,1,""]},"ovchipcardlib.ovchipcardlib":{Service:[6,4,1,""]},"ovchipcardlib.ovchipcardlib.Service":{authorization_token:[6,1,1,""],cards:[6,1,1,""],get_card_by_alias:[6,2,1,""]},"ovchipcardlib.ovchipcardlibexceptions":{BrokenResponse:[6,3,1,""],UnAuthorized:[6,3,1,""],UnableToRetrieveToken:[6,3,1,""]},ovchipcardlib:{helpers:[6,0,0,"-"],ovchipcardlib:[6,0,0,"-"],ovchipcardlibexceptions:[6,0,0,"-"]}},objnames:{"0":["py","module","Python module"],"1":["py","attribute","Python attribute"],"2":["py","method","Python method"],"3":["py","exception","Python exception"],"4":["py","class","Python class"],"5":["py","staticmethod","Python static method"]},objtypes:{"0":"py:module","1":"py:attribute","2":"py:method","3":"py:exception","4":"py:class","5":"py:staticmethod"},terms:{"__getnewargs__":6,"__getstate__":6,"__repr__":6,"__str__":6,"case":7,"class":6,"default":6,"import":[6,8],"public":7,"return":[6,7],"static":6,"true":8,"while":8,absolutelli:7,access:6,accord:7,account:7,action:7,actual:7,add:1,after:6,alia:6,all:[6,8],also:7,alwai:[1,7],ani:7,api:7,app:7,appar:7,appear:7,appreci:1,appropri:6,assum:1,auth_token:6,authent:[0,6,7],author:6,authorization_token:6,auto_reload:6,awar:7,balanc:6,balance_datetim:6,base:[0,6,7],befor:7,benchmark:7,between:7,bit:1,bodi:8,bound:7,branch:1,brokenrespons:6,bugfix:1,call:6,can:[1,6,7],card:[6,7,8],card_nam:8,certain:7,chang:1,check_in_info:6,check_in_text:6,checkin:7,checkout:1,chip:[6,7],chipkaart:8,clear:6,clone:1,close:7,code:8,com:[0,1,7],command:4,commit:1,compani:6,conclus:6,content:[3,5],copi:[1,6],costa:0,costastf:1,could:[6,8],creat:1,credenti:6,dai:7,data:6,datetim:[6,8],delai:[7,8],delay_messag:8,descript:1,design:7,detail:1,did:6,discard:8,drift:7,e_purse_mut:6,e_purse_mut_info:6,easier:1,enter:8,event:[7,8],everi:1,except:6,exclud:6,expect:[6,7],expires_in:6,expiry_datetim:6,explain:1,explan:6,extens:7,fail:6,fare:6,fare_calcul:6,fare_text:6,featur:[1,3],field:6,file:6,first:[0,2],follow:8,fork:1,format:[6,8],found:6,from:[6,8],get_card_by_alia:[6,8],get_latest_transact:[6,7,8],git:1,github:[0,1,7],given:1,gmail:0,got:8,greatli:1,guarante:7,handl:7,happen:6,have:[1,4],hefti:7,help:1,helper:5,here:[1,6,8],histor:8,how:1,http:[0,1,7],implement:[1,6],index:3,inform:7,instal:[1,3],instanti:6,interact:7,itself:6,job:7,join:8,keep:1,last:6,latest:[6,7],librari:7,line:4,list:[6,7],littl:1,live:6,local:1,log:8,look:6,loop:8,loos:[0,7],lot:7,main:6,make:1,match:6,merg:1,method:7,mijn:8,minut:7,mkvirtualenv:[1,4],mobil:7,model:6,modul:[3,5],monitor:[6,7,8],name:[1,6],narrow:1,netherland:7,nice:6,none:[0,6],now:[1,8],number:[6,8],object:6,old:8,onc:7,order:7,ordereddict:6,origin:1,out:7,ovchip:[0,7],ovchip_password:8,ovchip_usernam:8,ovchipapi:[0,7],ovchipcardlib:1,ovchipcardlibexcept:5,packag:[3,5],page:3,paramet:6,part:[0,6,7],password:6,pickl:6,pip:4,plain:6,possibl:1,present:6,print:8,prioriti:6,product_info:6,product_text:6,project:8,properli:7,propos:1,push:1,pypi:2,python:[0,1,7],rang:7,readi:1,readthedoc:7,real:7,refresh:6,releas:2,remot:6,repres:6,represent:6,request:1,respond:6,retriev:6,scope:[1,6],search:3,self:6,sequenc:7,server:[1,6],serversid:7,servic:[6,7,8],service_inst:6,set:1,setup:1,show:7,simpl:7,sinc:8,sleep:8,some:7,sort:7,sourc:6,station:6,statu:6,status_announc:6,str:[6,8],string:6,submodul:5,system:7,thei:[1,7],them:7,thi:[1,6],through:7,time:[6,7,8],time_shift:8,timestamp:6,timestamp_to_datetim:6,togeth:7,token:6,transact:[6,7,8],transport:7,tupl:6,tyf:0,tyfoxylo:0,type:6,unabletoretrievetoken:6,unauthor:6,upon:6,usag:3,usernam:6,util:7,veri:[0,7],verifi:6,virtualenv:1,virtualenvwrapp:[1,4],want:8,watch:8,web:7,webapi:7,welcom:1,why:0,work:1,would:[1,7],yet:0,you:[1,4,8],your:1},titles:["Credits","Contributing","History","Welcome to ovchipcardlib&#8217;s documentation!","Installation","ovchipcardlib","ovchipcardlib package","ovchipcardlib","Usage"],titleterms:{content:6,contribut:1,contributor:0,credit:0,develop:0,document:3,featur:7,feedback:1,get:1,helper:6,histori:2,indic:3,instal:4,lead:0,modul:6,ovchipcardlib:[3,5,6,7],ovchipcardlibexcept:6,packag:6,start:1,submit:1,submodul:6,tabl:3,usag:8,welcom:3}})

ovchipcardlib/ovchipcardlibexceptions.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# File: ovchipcardlibexceptions.py
#
# Copyright 2017 Costas Tyfoxylos
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to
#  deal in the Software without restriction, including without limitation the
#  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
#  sell copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
#

"""
Custom exception code for ovchipcardlib.

.. _Google Python Style Guide:
   http://google.github.io/styleguide/pyguide.html

"""

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = 'google'
__date__ = '''31-03-2017'''
__copyright__ = '''Copyright 2017, Costas Tyfoxylos'''
__credits__ = ["Costas Tyfoxylos"]
__license__ = '''MIT'''
__maintainer__ = '''Costas Tyfoxylos'''
__email__ = '''<costas.tyf@gmail.com>'''
__status__ = '''Development'''  # "Prototype", "Development", "Production".


class UnableToRetrieveToken(Exception):
    """The retrieval of the token failed."""


class UnAuthorized(Exception):
    """Could not verify authorization token."""


class BrokenResponse(Exception):
    """Server did not respond as expected."""


ovchipcardlib/ovchipcardlib.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# File: ovchipcardlib.py
#
# Copyright 2017 Costas Tyfoxylos
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to
#  deal in the Software without restriction, including without limitation the
#  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
#  sell copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
#

"""
Main code for ovchipcardlib.

.. _Google Python Style Guide:
   http://google.github.io/styleguide/pyguide.html

"""

import logging

import requests

from .helpers import Token, Card
from .ovchipcardlibexceptions import (UnableToRetrieveToken,
                                      UnAuthorized,
                                      BrokenResponse)

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = '''google'''
__date__ = '''14-04-2017'''
__copyright__ = '''Copyright 2017, Costas Tyfoxylos'''
__credits__ = ["Costas Tyfoxylos"]
__license__ = '''MIT'''
__maintainer__ = '''Costas Tyfoxylos'''
__email__ = '''<costas.tyf@gmail.com>'''
__status__ = '''Development'''  # "Prototype", "Development", "Production".


# This is the main prefix used for logging
LOGGER_BASENAME = '''ovchipcardlib'''
LOGGER = logging.getLogger(LOGGER_BASENAME)
LOGGER.addHandler(logging.NullHandler())


CLIENT_ID = 'nmOIiEJO5khvtLBK9xad3UkkS8Ua'
CLIENT_SECRET = 'FE8ef6bVBiyN0NeyUJ5VOWdelvQa'  # noqa


class Service:  # pylint: disable=too-many-instance-attributes
    """Object representing the ov chip card service.

    Can authenticate with the appropriate credentials and retrieves cards
    from the remote service.
    """

    def __init__(self, username, password):
        """Initializing of the Service object.

        Args:
            username (str): The username to log in with
            password (str): The password to log in with

        """
        logger_name = '{base}.{suffix}'.format(base=LOGGER_BASENAME,
                                               suffix=self.__class__.__name__)
        self._logger = logging.getLogger(logger_name)
        self.username = username
        self.password = password
        self._client_id = CLIENT_ID
        self._client_secret = CLIENT_SECRET
        self._auth_url = 'https://login.ov-chipkaart.nl/oauth2/token'
        self.api_url = 'https://api2.ov-chipkaart.nl/femobilegateway/v1'
        self.locale = 'nl-NL'
        self._cards = []
        self._auth_token = None
        self._token = self._get_token()

    def _get_token(self):
        """Gets the authentication token from the service.

        Returns:
            Token object of the retrieved token

        """
        params = {'username': self.username,
                  'password': self.password,
                  'client_id': self._client_id,
                  'client_secret': self._client_secret,
                  'grant_type': 'password',
                  'scope': 'openid'}
        return self._retrieve_token(params)

    def _refresh_token(self):
        """Refreshes the authentication token from the service.

        Returns:
            True on success

        """
        params = {'refresh_token': self._token.refresh,
                  'client_id': self._client_id,
                  'client_secret': self._client_secret,
                  'grant_type': 'refresh_token'}
        self._token = self._retrieve_token(params)
        return True

    def _retrieve_token(self, data):
        """Does the actual posting to the service for the token.

        Args:
            data (dict): The dictionary with the arguments to post to the
                service

        Returns:
            Token object of the retrieved token

        """
        response = requests.post(self._auth_url, data=data)
        if not response.ok:
            error = response.json().get('error_description', 'Unknown error!')
            raise UnableToRetrieveToken(error)
        token = response.json()
        return Token(token.get('access_token'),
                     token.get('expires_in'),
                     token.get('id_token'),
                     token.get('refresh_token'),
                     token.get('scope'),
                     token.get('token_type'))

    @property
    def authorization_token(self):
        """The authorization token.

        Returns:
             auth_token (str): The string of the authentication token

        """
        if not self._auth_token:
            params = {'authenticationToken': self._token.id}
            url = '{base}/api/authorize'.format(base=self.api_url)
            try:
                response = requests.post(url, data=params)
                data = response.json()
            except ValueError:
                message = ('Status code :{}\n'
                           'Response text :{}'.format(response.status_code,
                                                      response.text))
                raise BrokenResponse(message)
            status_code = data.get('c')
            exception = data.get('e')
            authorization = data.get('o')
            if not status_code == 200:
                raise UnAuthorized(exception or authorization)
            self._auth_token = authorization
        return self._auth_token

    @property
    def cards(self):
        """Retrieves the cards and instantiates the Card objects.

        Returns:
             cards (list): A list of card objects

        """
        if not self._cards:
            params = {'authorizationToken': self.authorization_token,
                      'locale': self.locale}
            url = '{base}/cards/list'.format(base=self.api_url)
            try:
                response = requests.post(url, data=params)
                data = response.json()
            except ValueError:
                # we take for granted that the response went through. This
                # will fail miserably in problematic network conditions where
                #  the response would not actually get through.
                message = ('Status code :{}\n'
                           'Response text :{}'.format(response.status_code,
                                                      response.text))
                raise BrokenResponse(message)
            status_code = data.get('c')
            exception = data.get('e')
            data = data.get('o')
            if not status_code == 200:
                raise LookupError(exception or data)
            self._cards = [Card(self, info) for info in data]
        return self._cards

    def get_card_by_alias(self, alias):
        """Returns a card object by alias.

        Args:
            alias (str):The alias of the card to look for

        Returns:
             A card object if a match found or None

        """
        return next((card for card in self.cards
                     if card.alias.lower() == alias.lower()), None)
